src/conn/conn_tiered.c

```diff
@@ -121,55 +131,58 @@ __flush_tier_once(WT_SESSION_IMPL *session, uint32_t flags)
+__tier_storage_remove_local(WT_SESSION_IMPL *session)
 {
-    WT_CONFIG_ITEM cval;
     WT_DECL_RET;
-    size_t len;
+    WT_TIERED_WORK_UNIT *entry;
     uint64_t now;
-    char *config, *newfile;
-    const char *cfg[2], *filename;
-
-    config = newfile = NULL;
-    if (uri == NULL)
-        return (0);
-    __wt_verbose(session, WT_VERB_TIERED, "Removing tree %s", uri);
-    filename = uri;
-    WT_PREFIX_SKIP_REQUIRED(session, filename, "tiered:");
-    len = strlen("file:") + strlen(filename) + 1;
-    WT_ERR(__wt_calloc_def(session, len, &newfile));
-    WT_ERR(__wt_snprintf(newfile, len, "file:%s", filename));
+    const char *object;
 
-    /*
-     * If the file:URI of the tiered object does not exist, there is nothing to do.
-     */
-    ret = __wt_metadata_search(session, newfile, &config);
-    if (ret == WT_NOTFOUND) {
-        ret = 0;
-        goto err;
-    }
-    WT_ERR(ret);
+    entry = NULL;
+    for (;;) {
+        /* Check if we're quitting or being reconfigured. */
+        if (!__tiered_server_run_chk(session))
+            break;
 
-    /*
-     * We have a local version of this tiered data. Check its metadata for when it expires and
-     * remove if necessary.
-     */
-    cfg[0] = config;
-    cfg[1] = NULL;
-    WT_ERR(__wt_config_gets(session, cfg, "local_retention", &cval));
-    __wt_seconds(session, &now);
-    if (force || (uint64_t)cval.val + S2C(session)->bstorage->retain_secs >= now)
+        __wt_seconds(session, &now);
+        __wt_tiered_get_drop_local(session, now, &entry);
+        if (entry == NULL)
+            break;
+        WT_ERR(__wt_tiered_name(
+          session, &entry->tiered->iface, entry->id, WT_TIERED_NAME_OBJECT, &object));
+        __wt_verbose(session, WT_VERB_TIERED, "REMOVE_LOCAL: %s at %" PRIu64, object, now);
+        WT_PREFIX_SKIP_REQUIRED(session, object, "object:");
         /*
-         * We want to remove the entry and the file. Probably do a schema_drop on the file:uri.
+         * If the handle is still open, it could still be in use for reading. In that case put the
+         * work unit back on the work queue and keep trying.
          */
-        __wt_verbose(session, WT_VERB_TIERED, "Would remove %s. Local retention expired", newfile);
-
+        if (__wt_handle_is_open(session, object)) {
+            __wt_verbose(session, WT_VERB_TIERED, "REMOVE_LOCAL: %s in USE, queue again", object);
+            WT_STAT_CONN_INCR(session, local_objects_inuse);
+            /*
+             * FIXME-WT-7470: If the object we want to remove is in use this is the place to call
+             * object sweep to clean up block->ofh file handles. Another alternative would be to try
+             * to sweep and then try the remove call below rather than pushing it back on the work
+             * queue. NOTE: Remove 'ofh' from s_string.ok when removing this comment.
+             */
+            __wt_tiered_push_work(session, entry);
```

Comment:
Since we're queueing this entry again, and this loops forever until the queue is empty, won't we keep seeing the same entry over and over?  I wonder if we shouldn't increase the timestamp in the entry?

Reply:
That is a good point. I was thinking the conflict was a transient thing but you're right. There isn't a lot of harm in caching it for longer.